File: 0813-all-paths-from-source-to-target/0813-all-paths-from-source-to-target.cpp
class Solution {
public:
    vector<vector<int>> allPathsSourceTargetDfs(vector<vector<int>>& graph) {
        int t = graph.size()-1;
        vector<vector<int>> ret;
        function<void(int, vector<int>)> dfs = [&](int u, vector<int> path) { 
            if(u==t) {
                ret.push_back(path);
            }
            for(int v:graph[u]) {
                path.push_back(v);
                dfs(v, path);
                path.pop_back();
            }
        };
        dfs(0,{0});
        return ret;
    }

    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        int size = graph.size()-1;
        vector<vector<int>> ret;
        deque<vector<int>> bfs(1,vector(1,0));
        while(!bfs.empty()) {
            vector<int> temp = bfs.front();
            bfs.pop_front();
            if(temp.back() == size) {
                ret.push_back(temp);
            }
            for(auto& nei:graph[temp.back()]) {
                temp.push_back(nei);
                bfs.push_back(temp);
                temp.pop_back();
            }
            
        }
        return ret;
    }
};

            // for nei in graph[path[-1]]:
            //     q.insert(0,path + [nei])

File: 0813-all-paths-from-source-to-target/0813-all-paths-from-source-to-target.py
class Solution:
    def allPathsSourceTarget_29_30(self, graph: List[List[int]]) -> List[List[int]]:
        destination = len(graph)
        routes = defaultdict(list)
        routes[0] = [[0]]
        i = 0
        for nodes in graph:
            print(f"i={i}, nodes={nodes}")
            print(routes)
            for node in nodes:
                print(f'node={node}, routes[{i}]={routes[i]}')
                for route in routes[i]:
                    temp = list(route)
                    temp.append(node)
                    routes[node].append(temp)
            i+=1

        return routes[destination-1]


    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        def dfs(u: int, t: int, path: List[int]) -> Generator:
            if u == t:
                yield path

            for v in graph[u]:
                yield from dfs(v, t, path + [v])

        graph = {
            u: neighbours
            for u, neighbours in enumerate(graph)
        }
        return dfs(0, len(graph) - 1, [0])



        